apollo-router/src/plugins/authentication.rs

```diff
@@ -0,0 +1,944 @@
+                            .as_ref()
+                            .map_err(|e| {
+                                tracing::error!(%e, "could not activate commercial features");
+                                e
+                            })
+                            .ok()?
+                            .clone();
+
+                        my_client
+                            .get(url)
+                            .header(ACCEPT, APPLICATION_JSON.essence_str())
+                            .header(CONTENT_TYPE, APPLICATION_JSON.essence_str())
+                            .timeout(DEFAULT_AUTHENTICATION_NETWORK_TIMEOUT)
+                            .send()
+                            .await
+                            .map_err(|e| {
+                                tracing::error!(%e, "could not get url");
+                                e
+                            })
+                            .ok()?
+                            .text()
+                            .await
+                            .map_err(|e| {
+                                tracing::error!(%e, "could not process url content");
+                                e
+                            })
+                            .ok()?
+                    };
+                    let jwks: JwkSet = serde_json::from_str(&data)
+                        .map_err(|e| {
+                            tracing::error!(%e, "could not create JWKS from url content");
+                            e
+                        })
+                        .ok()?;
+                    Some(jwks)
+                };
+                Box::pin(fut)
+            });
+        let deduplicator = Deduplicate::with_capacity(getter, 1);
+
+        Ok(AuthenticationPlugin {
+            configuration: init.config.experimental.jwt,
+            jwks: Arc::new(deduplicator),
+            jwks_url: url,
+        })
+    }
+
+    fn router_service(&self, service: router::BoxService) -> router::BoxService {
+        let request_full_config = self.configuration.clone();
+        let request_jwks = self.jwks.clone();
+        let request_jwks_url = self.jwks_url.clone();
+
+        fn authentication_service_span() -> impl Fn(&router::Request) -> tracing::Span + Clone {
+            move |_request: &router::Request| {
+                tracing::info_span!(
+                    AUTHENTICATION_SPAN_NAME,
+                    "authentication service" = stringify!(router::Request),
+                    "otel.kind" = "INTERNAL"
+                )
+            }
+        }
+
+        ServiceBuilder::new()
+            .instrument(authentication_service_span())
+            .checkpoint_async(move |request: router::Request| {
+                let my_config = request_full_config.clone();
+                let my_jwks = request_jwks.clone();
+                let my_jwks_url = request_jwks_url.clone();
+                const AUTHENTICATION_KIND: &str = "JWT";
+
+                async move {
```

Comment:
could this be extracted to a function?

Reply:
probably, but most of the other plugins using checkpoint_async are just written to use an async block.